### PR TITLE
Shutdown asynchronous cluster instances correctly at exit

### PR DESCRIPTION
**Pull request by zanieb**
When working with clusters that have `asynchronous=True`, the `atexit` shutdown hook is not awaiting `cluster.close`. This results in lingering clusters and the following warning

```
/Users/mz/src/distributed/distributed/deploy/spec.py:663: RuntimeWarning: coroutine 'wait_for' was never awaited
  cluster.close(timeout=10)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
/Users/mz/src/distributed/distributed/deploy/spec.py:663: RuntimeWarning: coroutine 'SpecCluster._close' was never awaited
  cluster.close(timeout=10)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

The solution here is to force a synchronous call to `cluster.close` despite the value of `cluster.asynchronous`. (Here, I presume that the atexit handler cannot be an asynchronous function). ~I discovered that `cluster.sync(..., asynchronous=False)` was not respected so there is a small tweak there as well.~

~I've marked this as a draft so I can determine what tests need to change before requesting review.~

- ~[ ] Closes #xxxx~
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by madkinsz** [comment]
It looks like https://github.com/dask/distributed/pull/5531 resolved the issue with `cluster.sync`; thanks @jcrist :) -- let me know if there's anything else what would make this easier to accept.

File: distributed/deploy/tests/test_cluster.py [hunk 1]
```diff
@@ -43,3 +45,44 @@ def __init__(self):
 
     cluster = FooCluster()
     assert "foo" in cluster._cluster_info
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize("asynchronous_cluster", [True, False])
+async def test_sync_defaults_to_cluster_setting(
```
Comment (fjetter) [inline distributed/deploy/tests/test_cluster.py:52]: There is still something wrong. This test fails reliably on CI and locally for me with

```
=================================== FAILURES ===================================
_________________ test_sync_defaults_to_cluster_setting[True] __________________

asynchronous_cluster = True
loop_in_thread = <tornado.platform.asyncio.AsyncIOLoop object at 0x7f3e749a8a10>
cleanup = None

    @pytest.mark.asyncio
    @pytest.mark.parametrize("asynchronous_cluster", [True, False])
    async def test_sync_defaults_to_cluster_setting(
        asynchronous_cluster, loop_in_thread, cleanup
    ):
    
        cluster = Cluster(asynchronous=asynchronous_cluster)
        cluster.loop = loop_in_thread
    
        async def foo():
            return 1
    
        result = cluster.sync(foo)
    
        if asynchronous_cluster:
>           assert isinstance(result, CoroutineType)
E           assert False
E            +  where False = isinstance(1, CoroutineType)

distributed/deploy/tests/test_cluster.py:65: AssertionError
```

Comment (madkinsz) [reply distributed/deploy/tests/test_cluster.py:52]: Ah sorry this looks like an issue after resolving the merge conflict. I'll investigate and get back to you. Thanks for taking a look!

**Comment by madkinsz** [comment]
Hey @jcrist — my test asserts that a `Cluster` instance with `asynchronous=True` will return a coroutine instead of a final result when `sync` is called with `asynchronous=None`. However, the `asynchronous` flag is ignored because some current loop inspection logic determines if `sync` should return a coroutine or not.

```python
    @property
    def asynchronous(self):
        """Are we running in the event loop?"""
        return in_async_call(self.loop, default=getattr(self, "_asynchronous", False))
```

```python
def in_async_call(loop, default=False):
    """Whether this call is currently within an async call"""
    try:
        return loop.asyncio_loop is asyncio.get_running_loop()
    except RuntimeError:
        # No *running* loop in thread. If the event loop isn't running, it
        # _could_ be started later in this thread though. Return the default.
        if not loop.asyncio_loop.is_running():
            return default
        return False
```

Modifying my test to toggle the loop instead of the `asynchronous` setting makes it pass

```python

@pytest.mark.asyncio
@pytest.mark.parametrize("use_current_loop", [True, False])
async def test_sync_defaults_to_checking_current_event_loop(
    cleanup, loop_in_thread, use_current_loop
):
    cluster = Cluster(loop=IOLoop.current() if use_current_loop else loop_in_thread)

    async def foo():
        return 1

    result = cluster.sync(foo)

    if use_current_loop:
        assert isinstance(result, CoroutineType)
        assert await result == 1
    else:
        assert result == 1
```

However, now we're back to getting cluster closure errors

```
sys:1: RuntimeWarning: coroutine 'Cluster._close' was never awaited
Task was destroyed but it is pending!
```

I'm not entirely sure what the intent is here.

**Comment by zanieb** [comment]
I presume this is still incorrect but is stale now.

**Pull request closed without merging**